src/coreclr/nativeaot/Runtime/RuntimeInstance.cpp

```diff
@@ -229,6 +237,65 @@ extern "C" bool __stdcall RegisterUnboxingStubs(PTR_VOID pvStartRange, uint32_t
     return GetRuntimeInstance()->RegisterUnboxingStubs(pvStartRange, cbRange);
 }
 
+#else // USE_PORTABLE_HELPERS
+
+bool RuntimeInstance::RegisterUnboxingStubTargetMappings(void* pMappings, uint32_t mappingsSizeInBytes)
+{
+    struct MappingsBlob
+    {
+        uint32_t SimpleUnboxingStubCount;
+        UnboxingStubTargetMapping Data[];
+    };
+
+    ASSERT(((mappingsSizeInBytes - sizeof(MappingsBlob)) % sizeof(UnboxingStubTargetMapping)) == 0);
+    auto initMap = [](UnboxingStubTargetsMap& map, UnboxingStubTargetMapping* mappings, uint32_t count) {
+        // Supporting multiple registrations would require synchronizing access to the tables.
+        ASSERT(map.GetCount() == 0);
+        map.Reallocate(count);
+
+        for (size_t i = 0; i < count; i++)
+        {
+            ASSERT(map.Lookup(mappings[i].GetKey()) == nullptr); // No duplicates.
+            if (!map.Add(&mappings[i]))
+            {
+                return false;
+            }
+        }
+
+        return true;
+    };
+
+    MappingsBlob* allMappings = (MappingsBlob*)pMappings;
+    uint32_t allMappingsCount = mappingsSizeInBytes / sizeof(UnboxingStubTargetMapping);
+    uint32_t simpleStubCount = allMappings->SimpleUnboxingStubCount;
+    uint32_t instantiatingStubCount = allMappingsCount - simpleStubCount;
+
+    if (!initMap(m_unboxingStubTargetsMap, &allMappings->Data[0], simpleStubCount) ||
+        !initMap(m_unboxingAndInstantiatingStubTargetsMap, &allMappings->Data[simpleStubCount], instantiatingStubCount))
```

Comment:
The core runtime should not need this mapping for anything. Is that correct?

Would it make more sense to initialize this map lazily and do the whole thing in C#, like we do that for other maps that are needed by reflection?

Reply:
> The core runtime should not need this mapping for anything. Is that correct?

Yes. It's only consumed at the CoreLib layer and above.

> Would it make more sense to initialize this map lazily and do the whole thing in C#, like we do that for other maps that are needed by reflection?

I considered it. The main problem is that the functionality is used by delegate equality, which seems low-level enough to be non-trivial to initialize robustly in managed code.

Additionally, the `SHash` closed hashing table implementation is very nicely fitted to this use case, a managed dictionary would use 3x+ more memory.

Reply:
Does it need to be used by the delegate equality in the first place? It looks like that it is a left-over from some flavor of multi-module mode that allowed multiple unboxing stubs to exist for given method. It should not be a thing in current native AOT. You can try to upstream removing it from delegate equality.

Even if it still needed for some reason, delegate equality should not be used by low-level runtime code where the lazy initialization would lead to problems.

A reverse map like this one is a common problem for reflection-like functionality. I think it would be fine to implement this reverse map in a similar way. We have a few of those. For example, look for `FunctionPointersToOffsets`. (This one is also more complicated than it needs to be due to support for multimodule flavors that we do not care about anymore.)

In fact, it looks like this reverse map was implemented in managed code at some point. There are unused `UnboxingAndInstantiatingStubMap` and `struct UnboxingAndInstantiatingStubMapEntry` that one would need to implement reverse maps like this.